### PR TITLE
Dynamic sidenav menus - take 3

### PR DESCRIPTION
**Pull request by mpelzsherman**
## Description
Currently, array of drupal system menus fed into graphql query is hardcoded. This pr adds logic to grab system menu names from graphql endpoint. This will enable new sites to function in DevShop (and all other environments) without having to push out a new release.

This is the 3rd attempt at a PR that works in all environments, including production builds, and meets our coding standards.

## Testing done
Run `yarn build:content --pull-drupal`, note that side navs still show up on VAMC pages.

## Acceptance criteria
- [x] Facility nav list not hard-coded
- [x] Side nav menu is still present

## Definition of done
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

### CODE REVIEW

File: src/site/stages/build/drupal/graphql/GetLatestPageById.graphql.js [hunk 1]
```diff
@@ -104,8 +105,9 @@ module.exports = `
     ${menuLinksQuery}
   }
 `;
+};
 
-const query = module.exports;
+const query = buildQuery();
```
Comment (U-DON) [inline src/site/stages/build/drupal/graphql/GetLatestPageById.graphql.js:110]:
```suggestion
const query = await buildQuery();
```

File: src/site/stages/build/drupal/metalsmith-drupal.js [hunk 2]
```diff
@@ -133,6 +133,46 @@ function pipeDrupalPagesIntoMetalsmith(contentData, files) {
...
+    const contentApi = getApiClient(buildOptions);
+    log('Pulling side nav menus from Drupal...');
+    await contentApi.getSideNavigations().then(res => {
+      sideNavs = res.data.sideNavMenus;
+
+      // Write them to .cache/{buildtype}/drupal/side-nav-menus.json
+      fs.ensureDirSync(buildOptions.cacheDirectory);
+      fs.emptyDirSync(path.dirname(sideNavFile));
+      fs.writeJsonSync(sideNavFile, sideNavs, { spaces: 2 });
+    });
```
Comment (U-DON) [inline src/site/stages/build/drupal/metalsmith-drupal.js:163]: Since we're already using `await`, maybe for consistency:

```suggestion
    const response = await contentApi.getSideNavigations();
    sideNavs = response.data.sideNavMenus;

    // Write them to .cache/{buildtype}/drupal/side-nav-menus.json
    fs.ensureDirSync(buildOptions.cacheDirectory);
    fs.emptyDirSync(path.dirname(sideNavFile));
    fs.writeJsonSync(sideNavFile, sideNavs, { spaces: 2 });
```

File: src/site/stages/build/drupal/graphql/GetLatestPageById.graphql.js (hunk 1 above)
Comment (mpelzsherman) [reply src/site/stages/build/drupal/graphql/GetLatestPageById.graphql.js:110]: `buildQuery` is already `await`-ing the call to `facilitySidebarQuery.compiledQuery`. There is no need to `await`here, and in fact doing so would cause a parse error.

Comment (U-DON) [reply src/site/stages/build/drupal/graphql/GetLatestPageById.graphql.js:110]: Since it's an `async` function now, it's going to return the string wrapped in a Promise.

If you want to handle `query` as a string and not as a string wrapped in a Promise, you should `await` here.

Comment (U-DON) [reply src/site/stages/build/drupal/graphql/GetLatestPageById.graphql.js:110]: I suppose the issue is then that that part is not wrapped in an `async` function itself, so you can't `await`.

Comment (U-DON) [reply src/site/stages/build/drupal/graphql/GetLatestPageById.graphql.js:110]: So as it is now, you'd get an error like `query.replace is not a function`.

What the fix will look like is for this module to return something like `query.then(q => q.replace(...)`.

`getQuery`, which requires this module, is already treated like an async function in some places. But as I mentioned in another comment, you'll want to `await` its results in the other places where it's being used (like `getLatestPageById` in`drupal/api.js`).

Comment (mpelzsherman) [reply src/site/stages/build/drupal/graphql/GetLatestPageById.graphql.js:110]: OK @U-DON - I've made the changes you suggested.

It looks like `getLatestPageById` is only used by the preview server. 
However, I do not see the side navs when I try to preview facility pages on master, so I think this is an existing bug.

**Comment by mpelzsherman** [comment]
[Bug](https://github.com/department-of-veterans-affairs/va.gov-team/issues/15710) created for missing side nav in preview server.

File: src/site/stages/build/options.js [hunk 3]
```diff
@@ -236,4 +236,6 @@ async function getOptions(commandLineOptions) {
   return options;
 }
 
-module.exports = getOptions;
+module.exports = {
+  getOptions,
+};
```
Comment (U-DON) [inline src/site/stages/build/options.js:241]: Was this change necessary for something?

**Review by U-DON — DISMISSED** [review]

File: src/site/stages/build/options.js (hunk 3 above)
Comment (mpelzsherman) [reply src/site/stages/build/options.js:241]: I found it necessary to get the `require`'s working properly.

Comment (U-DON) [reply src/site/stages/build/options.js:241]: Did they not work properly before? This is the only change in this file.

Comment (U-DON) [reply src/site/stages/build/options.js:241]: I'm able to build without the named export. Unless you've observed the `require`s not working properly somewhere else.

Comment (mpelzsherman) [reply src/site/stages/build/options.js:241]: It was needed for the preview server.

File: src/site/stages/build/drupal/metalsmith-drupal.js [hunk 4]
```diff
@@ -133,6 +133,50 @@ function pipeDrupalPagesIntoMetalsmith(contentData, files) {
...
 
+/**
+ * Uses Drupal content via a new GraphQL query or the cached result of a
+ * previous query. This is where the cache is saved.
+ *
+ * @param {Object} buildOptions
+ * @return {Object} - The result of the GraphQL query
+ */
+async function getSideNavsViaGraphQL(buildOptions = global.buildOptions) {
+  const { getOptions } = require('../options');
```
Comment (U-DON) [inline src/site/stages/build/drupal/metalsmith-drupal.js:144]: I just noticed this; does this need to be a dynamic `require`? Could tihs be put at the top of the module?

Comment (mpelzsherman) [reply src/site/stages/build/drupal/metalsmith-drupal.js:144]: I tried it and this was the only way I could get it working.

Comment (U-DON) [reply src/site/stages/build/drupal/metalsmith-drupal.js:144]: By that, do you mean the build would fail?

Comment (mpelzsherman) [reply src/site/stages/build/drupal/metalsmith-drupal.js:144]: When I tried to view a page using the preview server, it would throw an error.

File: src/site/stages/build/drupal/api.js [hunk 5]
```diff
@@ -167,20 +168,26 @@ function getDrupalClient(buildOptions, clientOptionsArg) {
       }
     },
 
-    getAllPages(onlyPublishedContent = true) {
+    async getAllPages(onlyPublishedContent = true) {
+      const allPagesQuery = await getQuery(queries.GET_ALL_PAGES);
```
Comment (rianfowler) [inline src/site/stages/build/drupal/api.js:172]: This looks fine. 

One recommendation I have would be to move the subQuery call here: 
```suggestion
      const { getSideNavsViaGraphQL } = require('../../metalsmith-drupal');
      ...
      const sideNavs = await getSideNavsViaGraphQL();
      const allPagesQuery = getQuery(queries.GET_ALL_PAGES, sideNavs);
```

This might remove the need to spread the async stuff around. The const `facilitySidebarQuery.compiledQuery()` could be replaced with something called `makeFacilitySidebarQuery(sidenavs)`. 

The idea is to push the async stuff as far up as possible and inject the result into the query generating functions- so there's only one part of the code that's dealing with "async" and then everything after is just synchronous.

**Review by rianfowler — APPROVED** [review]
This looks good. I made a refactoring recommendation but we don't have to do it as part of this change.

File: src/site/stages/build/drupal/api.js (hunk 5 above)
Comment (mpelzsherman) [reply src/site/stages/build/drupal/api.js:172]: Thanks @rianfowler. Since this is a pretty large change, I'll log it as a tech debt ticket.

Comment (mpelzsherman) [reply src/site/stages/build/drupal/api.js:172]: https://github.com/department-of-veterans-affairs/va.gov-team/issues/15816
